consensus/tendermint/core/state_dumper.go

```diff
@@ -0,0 +1,266 @@
+	Proposal    *common.Hash
+	LockedValue *common.Hash
+	LockedRound int64
+	ValidValue  *common.Hash
+	ValidRound  int64
+
+	// committee state
+	Committee       types.Committee
+	Proposer        common.Address
+	IsProposer      bool
+	QuorumVotePower uint64
+	RoundStates     []RoundState
+	ProposerPolicy  uint64
+
+	// extra state
+	SentProposal          bool
+	SentPrevote           bool
+	SentPrecommit         bool
+	SetValidRoundAndValue bool
+
+	// timer state
+	BlockPeriod           uint64
+	ProposeTimerStarted   bool
+	PrevoteTimerStarted   bool
+	PrecommitTimerStarted bool
+
+	// current height messages.
+	CurHeightMessages []*MsgForDump
+	// backlog msgs
+	BacklogMessages []*MsgForDump
+	// backlog unchecked msgs.
+	UncheckedMsgs []*MsgForDump
+	// Known msg of gossip.
+	KnownMsgHash []common.Hash
+}
+
+func (c *core) CoreState() TendermintState {
+	state := TendermintState{}
+	// send state dump request.
+	var e = coreStateRequestEvent{
+		stateChan: make(chan TendermintState),
+	}
+	go c.sendEvent(e)
+	// wait for response with timeout.
+	timeout := time.After(time.Second)
+	select {
+	case s := <-e.stateChan:
+		state = s
+	case <-timeout:
+		state.Code = -1
+		c.logger.Debug("Waiting for tendermint core state timed out", "elapsed", time.Second)
+	}
+
+	return state
+}
+
+// State Dump is handled in the main loop triggered by an event rather than using RLOCK mutex.
+func (c *core) handleStateDump(e coreStateRequestEvent) {
+	state := TendermintState{
+		Client:            c.address,
+		ProposerPolicy:    uint64(c.proposerPolicy),
+		BlockPeriod:       c.blockPeriod,
+		CurHeightMessages: msgForDump(c.GetCurrentHeightMessages()),
+		BacklogMessages:   getBacklogMsgs(c),
+		UncheckedMsgs:     getBacklogUncheckedMsgs(c),
+		// tendermint core state:
+		Height:      c.Height(),
+		Round:       c.Round(),
+		Step:        uint64(c.step),
+		Proposal:    getProposal(c, c.Round()),
+		LockedValue: getHash(c.lockedValue),
+		LockedRound: c.lockedRound,
+		ValidValue:  getHash(c.validValue),
+		ValidRound:  c.validRound,
+
+		// committee state
+		Committee:       c.committeeSet().Committee(),
+		Proposer:        c.committeeSet().GetProposer(c.Round()).Address,
+		IsProposer:      c.isProposer(),
+		QuorumVotePower: c.committeeSet().Quorum(),
+		RoundStates:     getRoundState(c),
+		// extra state
+		SentProposal:          c.sentProposal,
+		SentPrevote:           c.sentPrevote,
+		SentPrecommit:         c.sentPrecommit,
+		SetValidRoundAndValue: c.setValidRoundAndValue,
+		// timer state
+		ProposeTimerStarted:   c.proposeTimeout.timerStarted(),
+		PrevoteTimerStarted:   c.prevoteTimeout.timerStarted(),
+		PrecommitTimerStarted: c.precommitTimeout.timerStarted(),
+		// known msgs in case of gossiping.
+		KnownMsgHash: c.backend.KnownMsgHash(),
+		Code:         0,
+	}
+
+	// for none blocking send state.
+	select {
+	case e.stateChan <- state:
+		c.logger.Debug("core state msg sent.")
+	default:
+		c.logger.Debug("core state msg dropped.")
+	}
```

Comment:
this problematic because it introduces a race condition where without timeout expiry you would have different result depending on which go routine checks it select statement first (line 94 vs line 145)

```suggestion
      c.logger.Debug("sending core state msg")
	e.stateChan <- state
```